core/lib/src/rocket.rs

```diff
@@ -641,135 +785,76 @@ impl Rocket {
-        self.router = match self.router.collisions() {
-            Ok(router) => router,
-            Err(e) => return Err(LaunchError::new(LaunchErrorKind::Collision(e)))
-        };
-
-        if let Some(failures) = self.fairings.failures() {
-            return Err(LaunchError::new(LaunchErrorKind::FailedFairings(failures.to_vec())))
-        }
-
-        Ok(self)
-    }
-
-    // TODO.async: Solidify the Listener APIs and make this function public
-    async fn listen_on<L>(mut self, listener: L) -> Result<(), crate::error::Error>
-    where
-        L: Listener + Send + Unpin + 'static,
-        <L as Listener>::Connection: Send + Unpin + 'static,
-    {
-        self = self.prelaunch_check().map_err(crate::error::Error::Launch)?;
-
-        self.fairings.pretty_print_counts();
-
-        // Determine the address and port we actually binded to.
-        self.config.port = listener.local_addr().map(|a| a.port()).unwrap_or(0);
-
-        let proto = if self.config.tls.is_some() {
-            "https://"
-        } else {
-            "http://"
-        };
-
-        let full_addr = format!("{}:{}", self.config.address, self.config.port);
-
-        // Set the keep-alive.
-        // TODO.async: implement keep-alive in Listener
-        // let timeout = self.config.keep_alive.map(|s| Duration::from_secs(s as u64));
-        // listener.set_keepalive(timeout);
-
-        // Freeze managed state for synchronization-free accesses later.
-        self.state.freeze();
-
-        // Run the launch fairings.
-        self.fairings.handle_launch(&self);
-
-        launch_info!("{}{} {}{}",
-                     Paint::masked("🚀 "),
-                     Paint::default("Rocket has launched from").bold(),
-                     Paint::default(proto).bold().underline(),
-                     Paint::default(&full_addr).bold().underline());
-
-        // Restore the log level back to what it originally was.
-        logger::pop_max_level();
-
-        // We need to get this before moving `self` into an `Arc`.
-        let mut shutdown_receiver = self.shutdown_receiver
-            .take().expect("shutdown receiver has already been used");
-
-        let rocket = Arc::new(self);
-        let service = hyper::make_service_fn(move |connection: &<L as Listener>::Connection| {
-            let rocket = rocket.clone();
-            let remote_addr = connection.remote_addr().unwrap_or_else(|| ([0, 0, 0, 0], 0).into());
-            async move {
-                Ok::<_, std::convert::Infallible>(hyper::service_fn(move |req| {
-                    hyper_service_fn(rocket.clone(), remote_addr, req)
-                }))
-            }
-        });
-
-        #[derive(Clone)]
-        struct TokioExecutor;
-
-        impl<Fut> hyper::Executor<Fut> for TokioExecutor where Fut: Future + Send + 'static, Fut::Output: Send {
-            fn execute(&self, fut: Fut) {
-                tokio::spawn(fut);
+    pub(crate) fn finish(&mut self) -> BoxFuture<'_, ()> {
```

Comment:
Can this not be an `async fn`? Also, the name `finish`, to me, implies some finality, but we call this function multiple times. Maybe `forward` or `advance` or `fire` or something?

Reply:
I *believe* this is not an `async fn` in order to break recursion with `_attach` - if that's indeed the case I will make a note of that.

Agreed on naming and I will think on that some more.

Reply:
I'm not sure I follow. Why would making the function an `async fn` lead to recursion where it otherwise wouldn't? Also, any suggestions on a name change? Perhaps `process` or `process_manifest`?

Reply:
`actualize_manifest`, `execute_manifest`, `consume_manifest`?

Reply:
Of those three, `actualize_manifest` sounds the closest to what it actually does - I'll go with that one.